### ISSUE TITLE
Alter config lookup?

### ISSUE THREAD

**Issue by nzakas** [eslint/eslint#2736]
this was brought up in the chat room, and I want to make sure we don't lose the discussion.there still seems to be a lot of confusion around configuration look up and how far it does or should go up to the directory structure. We had up until this point I looked all the way up to the home directory for configuration files. We would only use a configuration file in the home directory if there were no other configuration files found in subdirectories.

the current way this works was implemented here: https://github.com/eslint/eslint/issues/963

there was a discussion about whether or not we should keep looking up the directory structure or if we should stop at the current working directory. We eventually resolved to go all the way up to the home directory: https://github.com/eslint/eslint/issues/939#issuecomment-44234512

The question now is whether or not that choice was correct, and if it was correct, do we want a way to allow people to opt-out and specify where they want the lookup to stop.

The suggestion in the chatroom was to add "root": true to a config to say " don't look up any further from here. "

What do people think?

**Comment by gyandeeps** [comment]
I personally would like the root option. :+1:

**Comment by btmills** [comment]
Short of having a full IDE-esque project system (which is not something we want), I think the `root` option is the closest we'll get to "this is the top of my project; don't let anything else clobber it." I'm in favor :+1:

**Comment by fakewaffle** [comment]
I've wasted a few hours in the past by having config lookup further than the directory it was in. Is there a way to prevent searching outside of the current folder?

**Comment by ilyavolodin** [comment]
@fakewaffle Not currently, that's exactly the feature that is being discussed in this thread.

**Comment by nzakas** [comment]
I'd like to hear some opinions on whether or not the decision to keep looking up past the current working directory was correct or not: should we just stop at the cwd always?

**Comment by ilyavolodin** [comment]
I can't come up with scenario where looking up past `cwd` would be required, but I'm sure somebody has a usecase.

**Comment by xjamundx** [comment]
> should we just stop at the cwd always?

sounds reasonable.

Not following why we'd need to have anything other than that other than loose parity with `require()` rules, but it seems counter-intuitive and worth revisiting if it causes confusion IMO!

**Comment by lo1tuma** [comment]
I don't have a strong feeling about this, like @ilyavolodin I’ve never been in a situation of having a config file outside the cwd.

**Comment by fakewaffle** [comment]
I would prefer not to have it search outside the `cwd`.

**Comment by gcochard** [comment]
FWIW I do have a config for a few of my projects in the directory above them. It's a convenience mostly so I wouldn't be upset if this went away given the confusion it is causing.

**Comment by nzakas** [comment]
@xjamundx see link to the discussion in my original description. There was a string lobby for this functionality.

Keep in mind, if we make a change, then that's it - we can't change it again after this point.

**Comment by btmills** [comment]
I don't use upper-level configs either, but I do want to throw out one more possibility to see what the response would be:

Right now, configs are merged all the way up, falling back to `~/.eslintrc` if nothing is found between `cwd` and `/`. What if, instead of merging, it takes exactly one of `./.eslintrc`, `../.eslintrc`, `../../.eslintrc`, ..., `/.eslintrc`, `~/.eslintrc`, whichever comes first? So a `.eslintrc` in `cwd` would never be clobbered by anything above, but if I run `eslint` from within `eslint/lib/rules`, it still picks up `eslint/.eslintrc`.

**Comment by nzakas** [comment]
@btmills feels a bit too magical to me

**Comment by ilyavolodin** [comment]
@btmills I don't think that's a good idea. A common usecase is to have a separate config for `/test` and `/src` but a shared config at `/`. Your suggestion would make it impossible to do that, other then to use `extends`.

**Comment by btmills** [comment]
@ilyavolodin I guess I should clarify that my suggestion would only apply to `cwd` and above, and merging would still go on as normal below. That said, given your example, you'd have to run from the parent of both... which is no different from removing traversal above `cwd` entirely, so I'm in favor of stopping at `cwd`.

**Comment by ilyavolodin** [comment]
@btmills Ah.. Ok, got it. In that case I agree with @nzakas It does feel too "magical" to have a different behavior above `cwd` vs. below. It would be hard to explain to users, and I can foresee a tone of people coming back with problems that would be hard to trace.

**Comment by kentcdodds** [comment]
Just to make sure, once eslint *finds* an `.eslintrc` file, it stops climbing the directory tree right? Meaning, it's not going to find all of the parent `.eslintrc` files and try to merge them together right?

I'm pretty sure that's what you mean, but here's my reasoning for why I suggest you don't merge all parents: I have githooks that ensure that you're not committing any eslint errors. If I have a root .eslintrc and someone wanting to contribute to my project has a root .eslintrc and they conflict, then that could result in one of us not being able to commit code.

It would definitely be nice to have a base local config. I think I'm in favor of this.

**Comment by btmills** [comment]
> Just to make sure, once eslint finds an .eslintrc file, it stops climbing the directory tree right? Meaning, it's not going to find all of the parent .eslintrc files and try to merge them together right?

@kentcdodds my idea was that it would only look for the first `.eslintrc` above `cwd` and stop when it finds one, but given @ilyavolodin's example, I'm not too thrilled about it. Personally I'd prefer the search for `.eslintrc` to top out at `cwd`, though it would still merge in subdirectories.

**Comment by kentcdodds** [comment]
@btmills that's fine with me. Subdirectories (should) be committed to source so we wouldn't have the problem I mentioned.

**Comment by rgrove** [comment]
I like the `root` option proposal, and I would prefer that ESLint _not_ stop climbing at the cwd (or ever, unless it encounters an `.eslintrc` with `root: true`). I often have an `.eslintrc` in a project root or in my home folder but run ESLint from a deep subdirectory. In my use case, the cwd is only ever a starting point for the config search.

**Comment by kentcdodds** [comment]
If we got a `root` option, then I'd be fine with eslint climbing, that would solve my use case.

**Comment by gcochard** [comment]
Yes, `root` sounds like the best option to me as well. It keeps the old behavior as-is, and still follows the principle of least surprise.

**Comment by nzakas** [comment]
OK, thanks for the discussion. I'm not a fan of `root:true`, but it sounds like it would work for everyone so we shall make it so.

**Comment by ljharb** [comment]
(reposted from https://github.com/eslint/eslint/issues/820#issuecomment-111556029) I'm running into trouble with the merging behavior - a module A has a newer version of eslint with a new rule. it's dep is module B, which has an older version, and an eslintrc that doesn't have the new rule. In module A, npm explore B npm run eslint (where npm run eslint eslints my files), errors, because B's eslint is walking upwards and grabbing the incompatible eslint file from project A.

Is there a way I can disable the merging behavior, so that the first one found wins?
-------
From reading the discussion, it sounds like this will be coming?

**Comment by nzakas** [comment]
@ljharb issues marked as accepted are on the roadmap, so yes

**Comment by gyandeeps** [comment]
I will work on this and have the PR ready in few days.

**Comment by nzakas** [comment]
Awesome, thanks!

**Comment by gyandeeps** [comment]
As per the discussion on the PR, we should the suggestion was to add the `root` option to `.eslintrc` file.
But I think it will not work out that way because we need this option before we can start looking for `.eslintrc` files, thats the whole purpose of this option. 
If this option is present inside `.eslintrc` then we are saying go allout looking for `.eslintrc` file on the first run because we don't have the value of the option and then after we read the first `.eslintrc` file then respect the next lookup based on the option value. which seems odd to me.

How I understood this that thei would be a `cli` and `CLIEngine` option so that once we start looking for `.eslintrc` files we can actually use the option value to change the lookup behavior.

Guys am I missing something here?

**Comment by nzakas** [comment]
> But I think it will not work out that way because we need this option before we can start looking for .eslintrc files, thats the whole purpose of this option. 

This is incorrect. All we need to do is see if the current config file has root:true, and if so, stop searching up the directory structure. Otherwise, it continues as always.

Basically, I think we can add a check right here: https://github.com/eslint/eslint/blob/master/lib/config.js#L228

**Comment by gyandeeps** [comment]
Again based on other disvussion on top, Myself and others (possible) were thinking root mean cwd.
ie if option root is true then dont look outside cwd.

CC @ilyavolodin @lo1tuma

**Comment by nzakas** [comment]
I think you are conflating two different options discussed in this thread. I put this on the original description:

> The suggestion in the chatroom was to add "root": true to a config to say " don't look up any further from here. "

That's what we landed on as the final approach.

@btmills was suggesting an alternative related to cwd: https://github.com/eslint/eslint/issues/2736#issuecomment-111235852, but we decided against that.

**Comment by ljharb** [comment]
So, in other words, the first `.eslintrc`/config file found with `root: true` would cause it to stop spidering up looking for config files?

**Comment by nzakas** [comment]
Yes, exactly

**Comment by IanVS** [comment]
But if a user has both an `.eslintrc` and eslint configuration in `package.json`, and one of those has `root: true`, I assume that they both should be used/merged, but not any config files found in parent directories. Correct?

**Comment by ljharb** [comment]
I'm pretty confused as to why anyone would have *both* an `.eslintrc` and eslint configs in `package.json`. What's the use case there?

**Comment by IanVS** [comment]
The [docs](http://eslint.org/docs/user-guide/configuring#configuration-cascading-and-hierarchy) specifically mention this as being possible:

> If there is an `.eslintrc` and a `package.json` file found in the same directory, both will be used, with the `.eslintrc` having the higher precendence.

I've taken `root: true` to mean "Do not look in any parent folders for eslint config files."

**Comment by nzakas** [comment]
@IanVS correct. `root` in either location stops traversing up the directory structure but both files in that directory will be used.
